src/http_ece.rs

```diff
@@ -100,145 +50,74 @@ impl<'a> HttpEce<'a> {
         if content.len() > 3052 {
             return Err(WebPushError::PayloadTooLarge);
         }
-
-        let private_key =
-            agreement::EphemeralPrivateKey::generate(&agreement::ECDH_P256, &self.rng)?;
-        let public_key = private_key.compute_public_key()?;
         let mut salt_bytes = [0u8; 16];
-
         self.rng.fill(&mut salt_bytes)?;
-        let peer_public_key = agreement::UnparsedPublicKey::new(&agreement::ECDH_P256, self.peer_public_key);
-
-        agreement::agree_ephemeral(
-            private_key,
-            &peer_public_key,
-            WebPushError::Unspecified,
-            |shared_secret| match self.encoding {
-                ContentEncoding::AesGcm => {
-                    let mut payload = vec![0; 3054];
-                    front_pad(content, &mut payload);
-
-                    self.aes_gcm(
-                        shared_secret,
-                        public_key.as_ref(),
-                        &salt_bytes,
-                        &mut payload,
-                    )?;
-
-                    Ok(WebPushPayload {
-                        content: payload.to_vec(),
-                        crypto_headers: self.generate_headers(public_key.as_ref(), &salt_bytes),
-                        content_encoding: "aesgcm",
-                    })
-                }
-                ContentEncoding::Aes128Gcm => Err(WebPushError::NotImplemented),
-            },
-        )
-    }
-
-    pub fn generate_headers(
-        &self,
-        public_key: &'a [u8],
-        salt: &'a [u8],
-    ) -> Vec<(&'static str, String)> {
-        let mut crypto_headers = Vec::new();
-
-        let mut crypto_key = format!("dh={}", base64::encode_config(public_key, URL_SAFE_NO_PAD));
-
-        if let Some(ref signature) = self.vapid_signature {
-            crypto_key = format!("{}; p256ecdsa={}", crypto_key, signature.auth_k);
-
-            let sig_s: String = signature.into();
-            crypto_headers.push(("Authorization", sig_s));
-        };
 
-        crypto_headers.push(("Crypto-Key", crypto_key));
-        crypto_headers.push((
-            "Encryption",
-            format!("salt={}", base64::encode_config(&salt, URL_SAFE_NO_PAD)),
-        ));
-
-        crypto_headers
+        match self.encoding {
+            ContentEncoding::AesGcm => {
+                let encrypted_block =
+                    encrypt_aesgcm(self.peer_public_key, self.peer_secret, &salt_bytes, content)
+                        .map_err(|_| WebPushError::InvalidCryptoKeys)?;
+                let payload = encrypted_block.ciphertext.to_owned();
```

Comment:
Usually not being picky about things like this, but... Here we have the ciphertext by value, and we clone it right after to an internal payload. Could we either de-structure the returned block or move the ciphertext out? Asking, because I know this is hot code here and doing an extra allocation is something we should avoid.

Reply:
Yes! Definitely. I think this part is mostly me trying to make the borrow checker happy to dig out other errors and forgetting to get back to it.